jetstream.go

```diff
@@ -0,0 +1,334 @@
+func PublishStreamTimeout(t time.Duration) PublishOption {
+	return func(opts *jsOpts) error {
+		opts.timeout = t
+		return nil
+	}
+}
+
+// PublishCtx sets an interrupt context for waiting on a stream to reply
+func PublishCtx(ctx context.Context) PublishOption {
+	return func(opts *jsOpts) error {
+		opts.ctx = ctx
+		return nil
+	}
+}
+
+// AckWaitDuration waits for confirmation from the JetStream server
+func AckWaitDuration(d time.Duration) AckOption {
+	return func(opts *jsOpts) error {
+		opts.timeout = d
+		return nil
+	}
+}
+
+func (m *Msg) jsAck(body []byte, opts ...AckOption) error {
+	if m.Reply == "" {
+		return ErrMsgNoReply
+	}
+
+	if m == nil || m.Sub == nil {
+		return ErrMsgNotBound
+	}
+
+	m.Sub.mu.Lock()
+	nc := m.Sub.conn
+	m.Sub.mu.Unlock()
+
+	var err error
+
+	aopts := newJsOpts()
+	for _, f := range opts {
+		err = f(aopts)
+		if err != nil {
+			return err
+		}
+	}
+
+	if aopts.timeout == 0 {
+		return m.Respond(body)
+	}
+
+	_, err = nc.Request(m.Reply, body, aopts.timeout)
+
+	return err
+}
+
+// Ack acknowledges a JetStream messages received from a Consumer, indicating the message
+// should not be received again later
+func (m *Msg) Ack(opts ...AckOption) error {
+	return m.jsAck(AckAck, opts...)
+}
+
+// Nak acknowledges a JetStream message received from a Consumer, indicating that the message
+// is not completely processed and should be sent again later
+func (m *Msg) Nak(opts ...AckOption) error {
+	return m.jsAck(AckNak, opts...)
+}
+
+// AckProgress acknowledges a Jetstream message received from a Consumer, indicating that work is
+// ongoing and further processing time is required equal to the configured AckWait of the Consumer
+func (m *Msg) AckProgress(opts ...AckOption) error {
+	return m.jsAck(AckProgress, opts...)
+}
+
+// AckNext performs an Ack() and request that the next message be sent to subject ib
+func (m *Msg) AckNext(ib string) error {
+	return m.RespondMsg(&Msg{Subject: m.Reply, Reply: ib, Data: AckNext})
+}
+
+// AckTerm acknowledges a message received from JetStream indicating the message will not be processed
+// and should not be sent to another consumer
+func (m *Msg) AckTerm(opts ...AckOption) error {
+	return m.jsAck(AckTerm, opts...)
+}
+
+// JetStreamPublishAck metadata received from JetStream when publishing messages
+type JetStreamPublishAck struct {
+	Stream   string `json:"stream"`
+	Sequence int    `json:"seq"`
+}
+
+// ParsePublishAck parses the publish acknowledgement sent by JetStream
+func ParsePublishAck(m []byte) (*JetStreamPublishAck, error) {
+	if bytes.HasPrefix([]byte("-ERR"), m) {
+		if len(m) > 7 {
+			return nil, fmt.Errorf(string(m[6 : len(m)-1]))
+		}
+
+		return nil, fmt.Errorf(string(m))
+	}
+
+	if !bytes.HasPrefix(m, []byte("+OK {")) {
+		return nil, fmt.Errorf("invalid JetStream Ack: %v", string(m))
+	}
+
+	ack := &JetStreamPublishAck{}
+	err := json.Unmarshal(m[3:], ack)
+	return ack, err
+}
+
+func isValidJSName(n string) bool {
+	if n == "" {
```

Comment:
maybe just make one return line with both conditionals.

Reply:
will do, but this reads a lot better than combining it - compiler combines it anyway